python/ray/data/dataset.py

```diff
@@ -1479,68 +1491,112 @@ def filter(
+                    "Use predicate expressions from ray.data.expressions instead. "
+                    "For example: from ray.data.expressions import col; "
+                    "ds.filter(expr=col('column_name') > 5)",
+                    DeprecationWarning,
+                    stacklevel=2,
+                )
+
+                from ray.data._internal.planner.plan_expression.expression_evaluator import (  # noqa: E501
+                    ExpressionEvaluator,
+                )
+
+                # TODO: (srinathk) bind the expression to the actual schema.
+                # If expr is a string, convert it to a pyarrow.dataset.Expression
+                # Initialize ExpressionEvaluator with valid columns, if available
+                # str -> Ray Data's Expression
+                resolved_expr = ExpressionEvaluator.get_ray_data_expression(expr)
 
-            compute = TaskPoolStrategy(size=concurrency)
+                compute = TaskPoolStrategy(size=concurrency)
+                filter_kwargs.update(
+                    {
+                        "predicate_expr": resolved_expr,
+                        "compute": compute,
+                    }
+                )
+            else:
+                # expr is an Expr object (predicate expression)
+                compute = TaskPoolStrategy(size=concurrency)
+                filter_kwargs.update(
+                    {
+                        "predicate_expr": expr,
+                        "compute": compute,
+                    }
+                )
         else:
             warnings.warn(
                 "Use 'expr' instead of 'fn' when possible for performant filters."
             )
 
-            if callable(fn):
-                compute = get_compute_strategy(
-                    fn=fn,
-                    fn_constructor_args=fn_constructor_args,
-                    compute=compute,
-                    concurrency=concurrency,
-                )
-            else:
+            if not callable(fn):
                 raise ValueError(
                     f"fn must be a UserDefinedFunction, but got "
                     f"{type(fn).__name__} instead."
                 )
 
-        ray_remote_args = merge_resources_to_ray_remote_args(
-            num_cpus,
-            num_gpus,
-            memory,
-            ray_remote_args,
-        )
+            compute = get_compute_strategy(
+                fn=fn,
+                fn_constructor_args=fn_constructor_args,
+                compute=compute,
+                concurrency=concurrency,
+            )
+
+            filter_kwargs.update(
+                {
+                    "fn": fn,
+                    "fn_args": fn_args,
+                    "fn_kwargs": fn_kwargs,
+                    "fn_constructor_args": fn_constructor_args,
+                    "fn_constructor_kwargs": fn_constructor_kwargs,
+                    "compute": compute,
+                }
+            )
```

Comment:
What i meant is rather use explicitly typed vars that will be set to null in some branch but not the other, rather than use kwargs